controller/object_store_controller.go

```diff
@@ -0,0 +1,1159 @@
+// |         ┌───────────────────────┐      │shutdown
+// |         │ PersistentVolume      ├──────┘
+// |         └───────────────────────┘
+//
+// From this ownership relationship and the mount dependencies, the order of
+// creation of the resources is determined.
+func (osc *ObjectStoreController) handleStarting(store *longhorn.ObjectStore) (err error) {
+	pvc, store, err := osc.getOrCreatePVC(store)
+	if err != nil {
+		return errors.Wrap(err, "API error while creating pvc")
+	}
+
+	vol, store, err := osc.getOrCreateVolume(store, pvc)
+	if err != nil {
+		return errors.Wrap(err, "API error while creating volume")
+	}
+
+	pv, store, err := osc.getOrCreatePV(store, vol)
+	if err != nil {
+		return errors.Wrap(err, "API error while creating volume")
+	}
+
+	dpl, store, err := osc.getOrCreateDeployment(store)
+	if err != nil {
+		return errors.Wrap(err, "API error while creating deployment")
+	}
+
+	_, store, err = osc.getOrCreateService(store)
+	if err != nil {
+		return errors.Wrap(err, "API error while creating service")
+	}
+
+	endpoints, store, err := osc.getOrCreateS3Endpoints(store)
+	if err != nil {
+		return errors.Wrap(err, "API error while creating S3 ingresses")
+	}
+	osc.logger.Infof("object store %v has  %v S3 endpoint(s)", store.Name, len(endpoints))
+	// if there are no public endpoints, add the implicit cluster-internal one
+	if len(store.Status.Endpoints) == 0 {
+		store.Status.Endpoints = append(store.Status.Endpoints, fmt.Sprintf("%v.%v.svc", store.Name, osc.namespace))
+	}
+
+	if err := osc.checkPVC(pvc); err != nil {
+		return nil
+	}
+
+	if err := osc.checkVolume(vol); err != nil {
+		return nil
+	}
+
+	if err := osc.checkPV(pv); err != nil {
+		return nil
+	}
+
+	if err := osc.checkDeployment(dpl, store); err != nil {
+		return nil
+	}
+
+	logrus.Infof("object store %v is now running", store.Name)
+	store.Status.State = longhorn.ObjectStoreStateRunning
+	return nil
+}
+
+// This function does a short sanity check on the various resources that are
+// needed to operate the object stores. If any of them is found to be
+// unhealthy, the controller will transition the object store to "Error"
+// state, otherwise do nothing.
+func (osc *ObjectStoreController) handleRunning(store *longhorn.ObjectStore) (err error) {
+	if store.Spec.TargetState == longhorn.ObjectStoreStateStopped {
+		logrus.Infof("object store %v is now stopping", store.Name)
+		store.Status.State = longhorn.ObjectStoreStateStopping
+		return nil
+	}
+
+	dpl, err := osc.ds.GetDeployment(store.Name)
+	if err != nil {
+		return errors.Wrapf(err, "failed to find deployment %v", store.Name)
+	} else if err = osc.checkDeployment(dpl, store); err != nil {
+		logrus.Errorf("Object Store running but deployment not ready")
+		store.Status.State = longhorn.ObjectStoreStateError
+		return err
+	}
+
+	_, err = osc.ds.GetService(osc.namespace, store.Name)
+	if err != nil {
+		return errors.Wrapf(err, "failed to find service %v", store.Name)
```

Comment:
Why does not the controller update `store.Status.State` to `longhorn.ObjectStoreStateError` here?

Reply:
Agree we should update `store.Status.State` to `longhorn.ObjectStoreStateError` if the error is ErrorNotFound. Similarly for other resources

Reply:
Looks like the updated code always update the store.Status.State to error when hitting any error. This could be problematic as the error could be just a temporary error. Can we update store.Status.State to longhorn.ObjectStoreStateError if the error is ErrorNotFound? Can we this for other resources as well?